packages/@lwc/engine-core/src/framework/template.ts

```diff
@@ -113,6 +115,38 @@ function validateLightDomTemplate(template: Template, vm: VM) {
     }
 }
 
+export function parseFragment(strings: string[], ...keys: number[]): () => Element {
+    return function (): Element {
+        const {
+            context: { hasScopedStyles, stylesheetToken },
+            shadowMode,
+        } = getVMBeingRendered()!;
+        const classToken = hasScopedStyles && stylesheetToken ? ' ' + stylesheetToken : '';
+        const classAttrToken =
+            hasScopedStyles && stylesheetToken ? ` class="${stylesheetToken}"` : '';
+        const attrToken =
+            stylesheetToken && shadowMode === ShadowMode.Synthetic ? ' ' + stylesheetToken : '';
```

Comment:
Unfortunately it looks like this doesn't work if the same component is rendered as both synthetic and native on the same page ([demo](https://github.com/nolanlawson/lwc-barebone/commit/447211ca80688c4954679cbe95354c1267f5ff1a)). If it's rendered as native first, then the synthetic one won't get the styles:

![Screen Shot 2022-04-12 at 12 15 03 PM](https://user-images.githubusercontent.com/283842/163037073-c55480f8-c218-45fa-ab1b-260a4649f9d4.png)

The problem is that the second one doesn't get the scope token attribute. Wheres if you switch the order, it works because both get the attribute:

![Screen Shot 2022-04-12 at 12 18 06 PM](https://user-images.githubusercontent.com/283842/163037223-10b6a602-8bb2-48f3-acb6-1b3708f1fa6c.png)

I can think of two possible solutions:

1. Add the attribute every time, even in native (a small perf hit; see data in https://github.com/salesforce/lwc/pull/2670)
2. Move the caching from the component `ctx` to the `parseFragment` function itself. For instance, the anonymous function could cache based on whether `stylesheetToken` has changed.

My personal preference is for # 2 since it would not cause even a small perf regression.

Reply:
niiice catch! resolved in https://github.com/salesforce/lwc/pull/2781/commits/dcc5687a74dc5669ad5a11af41ea17439dfa6832